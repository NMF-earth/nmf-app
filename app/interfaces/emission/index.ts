import {
  Emission,
  EmissionEnum,
  EmissionPurchaseEnum
} from "./emission.interface";

export {
  Emission,
  EmissionEnum,
  EmissionPurchaseEnum
};
