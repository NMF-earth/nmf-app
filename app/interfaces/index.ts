import {
  Emission,
  EmissionEnum,
  EmissionPurchaseEnum
} from "./emission";

export {
  Emission,
  EmissionEnum,
  EmissionPurchaseEnum
};
