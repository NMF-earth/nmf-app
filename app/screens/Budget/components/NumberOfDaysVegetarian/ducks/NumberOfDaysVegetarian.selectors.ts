import { pipe, propOr, includes, filter, isEmpty, not } from "ramda";
import moment from "moment";
import { FoodEnum } from "carbon-footprint";
import { emissions } from "../../../../../ducks";
import { calculation } from "../../../../../utils";

const meatArray = [
  FoodEnum.beef,
  FoodEnum.chicken,
  FoodEnum.fish,
  FoodEnum.lamb,
  FoodEnum.pork,
  FoodEnum.redMeat,
  FoodEnum.tuna,
  FoodEnum.turkey,
  FoodEnum.whiteMeat
];

const isMeatEmission = emission =>
  includes(emission.emissionModelType, meatArray);

const getCreationDate = propOr(
  moment()
    .utc()
    .toISOString(),
  "creationDate"
);
const getDaysElapsedSinceToday = date => moment().diff(date, "days");

const getFoodEmissions = emissions.selectors.getFoodEmissions;
const getMeatEmissions = pipe(getFoodEmissions, filter(isMeatEmission));
const isAnyMeatEmissionSaved = pipe(getMeatEmissions, isEmpty, not);

const getLatestEmission = calculation.getLatestEmission;
const getDaysSinceEmission = emission =>
  pipe(getCreationDate, moment, getDaysElapsedSinceToday)(emission);

const getDaysWithoutEatingMeat = state =>
  pipe(getMeatEmissions, getLatestEmission, getDaysSinceEmission)(state);

export default {
  isAnyMeatEmissionSaved,
  getDaysWithoutEatingMeat
};
