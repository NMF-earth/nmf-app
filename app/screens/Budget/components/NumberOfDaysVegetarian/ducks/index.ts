import selectors from "./NumberOfDaysVegetarian.selectors";

export { selectors };
