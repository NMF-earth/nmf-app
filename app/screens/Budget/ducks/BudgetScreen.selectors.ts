import { pipe, filter, map, sum } from "ramda";
import moment from "moment";

import { budget, emissions } from "ducks";

import { calculation } from "../../../utils";

const isEmissionInCurrentMonth = (emission) =>
  moment(emission.creationDate).isSame(new Date(), "month") &&
  moment(emission.creationDate).isSame(new Date(), "year");
const isEmissionInCurrentYear = (emission) =>
  moment(emission.creationDate).isSame(new Date(), "year");

const getCarbonValue = pipe(
  map(calculation.getC02ValueFromEmission),
  sum,
  Math.round
);

const getCurrentMonthEmissions = filter(isEmissionInCurrentMonth);

const getCurrentMonthTransportCarbonValue = pipe(
  emissions.selectors.getTransportEmissions,
  getCurrentMonthEmissions,
  getCarbonValue
);
const getCurrentMonthFoodCarbonValue = pipe(
  emissions.selectors.getFoodEmissions,
  getCurrentMonthEmissions,
  getCarbonValue
);
const getCurrentMonthOtherCarbonValue = pipe(
  emissions.selectors.getOtherEmissions,
  getCurrentMonthEmissions,
  getCarbonValue
);

const getCurrentYearEmissions = filter(isEmissionInCurrentYear);

const getCurrentYearTransportCarbonValue = pipe(
  emissions.selectors.getTransportEmissions,
  getCurrentYearEmissions,
  getCarbonValue
);
const getCurrentYearFoodCarbonValue = pipe(
  emissions.selectors.getFoodEmissions,
  getCurrentYearEmissions,
  getCarbonValue
);
const getCurrentYearOtherCarbonValue = pipe(
  emissions.selectors.getOtherEmissions,
  getCurrentYearEmissions,
  getCarbonValue
);

const getCurrentMonthAllCarbonValue = (state) =>
  getCurrentMonthTransportCarbonValue(state) +
  getCurrentMonthFoodCarbonValue(state) +
  getCurrentMonthOtherCarbonValue(state);

const getCurrentYearAllCarbonValue = (state) =>
  getCurrentYearTransportCarbonValue(state) +
  getCurrentYearFoodCarbonValue(state) +
  getCurrentYearOtherCarbonValue(state);

export default {
  getMonthlyCarbonBudget: budget.selectors.getMonthlyCarbonBudget,
  getCurrentMonthTransportCarbonValue,
  getCurrentMonthFoodCarbonValue,
  getCurrentMonthOtherCarbonValue,
  getCurrentMonthAllCarbonValue,
  getCurrentYearTransportCarbonValue,
  getCurrentYearFoodCarbonValue,
  getCurrentYearOtherCarbonValue,
  getCurrentYearAllCarbonValue,
};
