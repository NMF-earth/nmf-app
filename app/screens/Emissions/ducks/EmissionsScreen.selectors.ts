import { map, pipe, groupBy, toPairs, sum } from "ramda";
import moment from "moment";

import { emissions } from "ducks";

import { Emission } from "../../../interfaces";
import { calculation, ui } from "../../../utils";

interface EmissionListItem extends Emission {
  title: string;
  co2value: number;
  iconName: string;
  onPress: () => void;
}

const getEmissionListItem = (item: Emission) => {
  const emissionItem: EmissionListItem = {
    ...item,
    title: ui.getTranslationModelType(item.emissionModelType),
    co2value: calculation.getC02ValueFromEmission(item),
    iconName: ui.getIconFromModelType(item.emissionModelType),
    onPress: () => {
      // do nothing.
    },
  };

  return emissionItem;
};

const getStartOfMonth = (time) => moment(time).startOf("month").format();

const groupByMonth = groupBy((item: EmissionListItem) =>
  getStartOfMonth(item.creationDate)
);

const dateObjMap = map(([date, data, co2value]) => ({
  date: date,
  data: data,
  co2value: co2value,
}));

/* moment().utc().toISOString() gives "YYYY-MM-DDTHH:mm:ss.sssZ" */
const filterByMostRecent = (array: [EmissionListItem]) =>
  array.sort((a, b) => +new Date(b.creationDate) - +new Date(a.creationDate));

const getMonthlyPourcentage = (items) =>
  map(
    (item) => [...item, sum(map((emission) => emission.co2value, item[1]))],
    items
  );

const getEmissions = (state) =>
  pipe(
    emissions.selectors.getAllEmissions,
    map(getEmissionListItem),
    filterByMostRecent,
    groupByMonth,
    toPairs,
    getMonthlyPourcentage,
    dateObjMap
  )(state);

export default {
  getEmissions,
  getEmissionListItem,
};
