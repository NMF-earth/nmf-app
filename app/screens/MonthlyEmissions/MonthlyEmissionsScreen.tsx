import React from "react";
import { FlatList, View } from "react-native";
import { useNavigation, useRoute } from "@react-navigation/native";
import { useSelector } from "react-redux";
import { map, pipe, propOr, sum } from "ramda";

import { navigate } from "navigation";
import { EmissionListItem, EmissionListItemProps, Text } from "components";
import { budget, userPreferences } from "ducks";
import { t, calculation } from "utils";
import { NavStatelessComponent } from "interfaces";

import { selectors } from "./ducks";
import styles from "./MonthlyEmissionsScreen.styles";
import navigationOptions from "./MonthlyEmissionsScreen.navigationOptions";

const getCO2value = map(propOr(0, "co2value"));
const getAmountCO2 = pipe(getCO2value, sum);

const MonthlyEmissions: NavStatelessComponent = () => {
  const route = useRoute();
  const monthlyCarbonBudget = useSelector(budget.selectors.getMonthlyCarbonBudget);

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const date = route.params?.date;
  const emissions: Array<EmissionListItemProps> = useSelector((state) =>
    selectors.getEmissions(state, date)
  );

  const navigation = useNavigation();
  const navigator = navigate(navigation);

  let percentageBudget = 1;
  const co2value = getAmountCO2(emissions);

  if (monthlyCarbonBudget && co2value) {
    percentageBudget = Math.round((co2value / monthlyCarbonBudget) * 100);
    if (percentageBudget < 1) {
      percentageBudget = 1;
    }
  }

  const useMetricUnits = useSelector(userPreferences.selectors.getUseMetricUnits);
  const getDisplayUnitsValue = calculation.getDisplayUnitsValue;
  const getDisplayUnits = calculation.getDisplayUnits;

  const renderHeader = () => (
    <View style={styles.containerHeader}>
      <Text.Primary darkGray bold>
        {`${getDisplayUnitsValue(co2value, useMetricUnits).toFixed(2)} ${getDisplayUnits(
          co2value,
          useMetricUnits
        )}CO2eq`}
      </Text.Primary>
      {percentageBudget && (
        <Text.Secondary orange={percentageBudget > 100} green={percentageBudget < 100}>
          {percentageBudget + " % " + t("MONTHLY_EMISSIONS_SCREEN_OF_BUDGET")}
        </Text.Secondary>
      )}
    </View>
  );

  return (
    <FlatList<EmissionListItemProps>
      style={styles.container}
      data={emissions}
      ListHeaderComponent={renderHeader()}
      keyExtractor={({ id }) => id}
      renderItem={({
        item: { id, isMitigated, name, title, co2value, iconName, emissionModelType },
      }) => (
        <EmissionListItem
          id={id}
          isMitigated={isMitigated}
          name={name}
          onPress={() => navigator.openEmissionItem({ id })}
          title={title}
          co2value={co2value}
          iconName={iconName}
          emissionModelType={emissionModelType}
        />
      )}
    />
  );
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
MonthlyEmissions.navigationOptions = navigationOptions;

export default MonthlyEmissions;
