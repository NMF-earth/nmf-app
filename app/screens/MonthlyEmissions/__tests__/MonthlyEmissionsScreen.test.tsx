import "react-native";
import React from "react";
import { create } from "react-test-renderer";
import { FoodType, TransportType } from "carbon-footprint";
import * as reactRedux from "react-redux";

import { emissions, userPreferences } from "ducks";
import { Emission, EmissionType } from "interfaces";

import { selectors } from "../ducks";
import MonthlyEmissions from "../MonthlyEmissionsScreen";

const emissionNotMitigatedOld: Emission = {
  id: "3",
  creationDate: new Date("2020-01-01T12:08:16.623Z").toISOString(),
  emissionModelType: FoodType.beans,
  emissionType: EmissionType.food,
  isMitigated: false,
  value: 30,
};

const emissionNotMitigated: Emission = {
  id: "1",
  creationDate: new Date("2020-03-01T12:08:16.623Z").toISOString(),
  emissionModelType: FoodType.beans,
  emissionType: EmissionType.food,
  isMitigated: false,
  value: 10,
};

const emissionMitigated: Emission = {
  id: "12",
  creationDate: new Date("2020-03-01T12:08:16.623Z").toISOString(),
  emissionModelType: TransportType.boat,
  emissionType: EmissionType.transport,
  isMitigated: true,
  value: 100,
};

const state = {
  [emissions.namespace]: [emissionNotMitigated, emissionMitigated, emissionNotMitigatedOld],
};

describe("<MonthlyEmissions />", () => {
  const useSelectorMock = jest.spyOn(reactRedux, "useSelector");

  beforeEach(() => {
    useSelectorMock.mockClear();
    useSelectorMock.mockImplementation(userPreferences.selectors.getUseMetricUnits);
  });

  test("render", () => {
    useSelectorMock
      .mockImplementationOnce(() => 200)
      .mockImplementationOnce(() => [
        ...selectors.getEmissions(state, new Date("2020-03-01T12:08:16.623Z").toISOString()),
      ]);

    const wrapper = create(<MonthlyEmissions />);
    expect(wrapper).toMatchSnapshot();
  });
});
