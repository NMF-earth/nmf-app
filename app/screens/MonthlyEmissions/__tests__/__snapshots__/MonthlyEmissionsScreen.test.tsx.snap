// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MonthlyEmissions /> render 1`] = `
<RCTScrollView
  ListHeaderComponent={
    <View
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomColor": "#DCF0E5",
          "borderBottomWidth": 2,
          "elevation": 0,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 16,
          "paddingVertical": 16,
          "shadowColor": "transparent",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowRadius": 0,
        }
      }
    >
      <Text.Primary
        bold={true}
        darkGray={true}
      >
        20.08 kgCO2eq
      </Text.Primary>
      <Text.Secondary
        green={true}
        orange={false}
      >
        10 % of budget
      </Text.Secondary>
    </View>
  }
  data={
    Array [
      Object {
        "co2value": 20,
        "creationDate": "2020-03-01T12:08:16.623Z",
        "emissionModelType": "beans",
        "emissionType": "food",
        "iconName": "md-nutrition",
        "id": "1",
        "isMitigated": false,
        "onPress": [Function],
        "title": "Beans",
        "value": 10,
      },
      Object {
        "co2value": 0.0803,
        "creationDate": "2020-03-01T12:08:16.623Z",
        "emissionModelType": "boat",
        "emissionType": "transport",
        "iconName": "md-boat",
        "id": "12",
        "isMitigated": true,
        "onPress": [Function],
        "title": "Boat",
        "value": 100,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
    }
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderBottomColor": "#DCF0E5",
            "borderBottomWidth": 2,
            "elevation": 0,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 16,
            "paddingVertical": 16,
            "shadowColor": "transparent",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowRadius": 0,
          }
        }
      >
        <Text.Primary
          bold={true}
          darkGray={true}
        >
          20.08 kgCO2eq
        </Text.Primary>
        <Text.Secondary
          green={true}
          orange={false}
        >
          10 % of budget
        </Text.Secondary>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={20}
        emissionModelType="beans"
        iconName="md-nutrition"
        id="1"
        isMitigated={false}
        onPress={[Function]}
        title="Beans"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={0.0803}
        emissionModelType="boat"
        iconName="md-boat"
        id="12"
        isMitigated={true}
        onPress={[Function]}
        title="Boat"
      />
    </View>
  </View>
</RCTScrollView>
`;
